slides/D-trait-objects-patterns.md

```diff
@@ -0,0 +1,824 @@
+        self.insert_money()
+    }
+}
+
+impl CoffeeMachine<MoneyInserted> {
+    fn make_beverage(self) -> CoffeeMachine<Idle> {
+        println!("There you go!");
+        self.into_state()
+    }
+}
+```
+</div>
+
+---
+layout: default
+---
+
+# Typestate: when to use
+
+- If your problem is like a state machine
+- Ensure *at compile* time that no invalid operation is done
+
+---
+layout: statement
+---
+
+# 4. The Strategy pattern
+Select behavior dynamically
+
+---
+layout: default
+---
+
+# Strategy: introduction
+
+- Turn set of behaviors into objects
+- Make them interchangeble inside context
+- Execute strategy depending on input
+
+*Trait objects work well here!*
+
+---
+layout: two-cols
+---
+
+# Strategy: example
+
+```rust
+
+trait PaymentStrategy {
+    fn pay(&self);
+}
+
+struct CashPayment;
+impl PaymentStrategy for CashPayment {
+    fn pay(&self) {
+        println!("🪙💸");
+    }
+}
+
+struct CardPayment;
+impl PaymentStrategy for CardPayment {
+    fn pay(&self) {
+        println!("💳");
+    }
+}
+```
+::right::
+
+<div style="padding-left:10px; padding-top: 50px;">
+
+```rust
+
+fn main() {
+    let method: &str 
+        = todo!("Read input");
+    let strategy: &dyn PaymentStrategy 
+        = match method {
+        "card" => &CardPayment,
+        "cash" => &CashPayment,
+        _ => panic!("Oh no!"),
+    };
+    strategy.pay();
+}
+```
+
+</div> 
+
+---
+layout: section
+---
+
+# The Deref polymorphism anti-pattern
+
+A common pitfall you'll want to avoid
+
+---
+layout: two-cols
+---
+
+# Deref polymorphism: Example
+
+```rust
+use std::ops::Deref;
+
+struct Animal {
+    name: String,
+}
+
+impl Animal {
+    fn walk(&self) {
+        println!("Tippy tap")
+    }
+    fn eat(&self) {
+        println!("Om nom")
+    }
+    fn say_name(&self) {
+        // Animals generally can't speak
+        println!("...")
+    }
+}
+```
+::right::
+```rust
+struct Dog {
+    animal: Animal
+}
+impl Dog {
+    fn eat(&self) {
+        println!("Munch munch");
+    }
+    fn bark(&self) {
+        println!("Woof woof!");
+    }
+}
+
+impl Deref for Dog {
+    type Target = Animal;
+
+    fn deref(&self) -> &Self::Target {
+        &self.animal
+    }
+}
+
+fn main (){ 
+    let dog: Dog = todo!("Instantiate Dog");
+    dog.bark();
+    dog.walk();
+    dog.eat();
+    dog.say_name();
+}
+```
+
+---
+layout: default
+---
+
+# The output
+
+```txt
+Woof woof!
+Tippy tap
+Munch munch
+```
+
+*Even overloading works!*
+
+---
+layout: default
+---
+
+# Why is it bad?
+
+- This is no 'real' inheritance: `Dog` is no subtype of `Animal`
+- Traits implemented on `Animal` are not implemented on `Dog` automatically
+- `Deref` and `DerefMut` are intended 'pointer-to-`T`' to `T` conversions
+- Deref coercion by `.` 'converts' `self` from `Dog` to `Animal`
+- Rust favours explicit conversions for easier reasoning about code
+
+*It will only add confusion: for OOP programmers it's incomplete, for Rust programmers it goes against semantics*
```

Comment:
In my ears, if something goes against the *semantics*, that means "this doesn't mean what you think it means, it prints 5 instead of 2". Here the code shown is a different beast; it's more that it goes against the "pragmatics": yes you can do this, but it will create more problems than it solves. Suggestion:

```suggestion
*It will only add confusion: for OOP programmers it's incomplete, for Rust programmers it is unidiomatic*
```